YatimaStdLib/SparseMatrix.lean

```diff
@@ -0,0 +1,193 @@
+  rows : Nat
+  cols : Nat
+
+instance [ToString R] : ToString (SparseMatrix R) where
+  toString m := toString ∘ Std.RBMap.Values.toArray $ m.entries
+
+open Std.RBMap
+
+/--
+Empty matrix
+-/
+def SparseMatrix.empty (col : Nat) (row : Nat) : SparseMatrix R :=
+  SparseMatrix.mk Std.RBMap.empty col row
+
+/--
+Insert a triple
+-/
+def SparseMatrix.insert
+  (col : Nat) (row : Nat) (a : R) (m : SparseMatrix R) : SparseMatrix R :=
+  SparseMatrix.mk
+    (Std.RBMap.insert m.entries (col,row) a)
+    m.rows
+    m.cols
+
+/--
+Creates a sparse matrix from a list
+-/
+def ofListWithDims
+  (rows : Nat) (cols : Nat) (l : List ((Nat × Nat) × R)) : SparseMatrix R :=
+  SparseMatrix.mk
+    (Std.RBMap.ofList l compare)
+    rows
+    cols
+
+/--
+Creates a sparse matrix from an array
+-/
+def ofArrayWithDims
+  (rows : Nat) (cols : Nat) (l : Array ((Nat × Nat) × R)) : SparseMatrix R :=
+  SparseMatrix.mk
+    (Std.RBMap.ofArray l compare)
+    rows
+    cols
+
+instance : Functor SparseMatrix where
+  map f m :=
+  SparseMatrix.mk
+    (Std.RBMap.mapVal (fun _ x => f x) m.entries)
+    m.rows
+    m.cols
+
+abbrev Row (R : Type) := Array R
+abbrev Col (R : Type) := Array R
+
+instance : Inhabited (SparseMatrix R) where
+  default := SparseMatrix.empty 0 0
+
+/--
+Matrix dimension
+-/
+def SparseMatrix.dim (m : SparseMatrix R) : Nat × Nat :=
+  (rows m, cols m)
+
+/--
+Returns a particular row by an index
+-/
+def SparseMatrix.row (m : SparseMatrix R) (i : Nat) : Row R :=
+  Std.RBMap.valuesArray $
+    Std.RBMap.mapKeys
+      (Std.RBMap.filter m.entries (fun j _ => i == j.1))
+      (fun (a,_) => a)
+
+/--
+Returns a particular column by an index
+-/
+def SparseMatrix.col (m : SparseMatrix R) (i : Nat) : Col R :=
+  Std.RBMap.valuesArray $
+    Std.RBMap.mapKeys
+      (Std.RBMap.filter m.entries (fun j _ => i == j.2))
+      (fun (_,b) => b)
+
+/--
+Transpose a sparse matrix
+-/
+def SparseMatrix.transpose (m : SparseMatrix R) : SparseMatrix R :=
+  SparseMatrix.mk
+    (Std.RBMap.mapKeys m.entries $ fun (a,b) => (b,a))
+    m.cols
+    m.rows
+
+variable {R : Type} [Ring R] [Coe Nat R]
+
+/--
+Returns a triple by given indices
+-/
+def SparseMatrix.findEntry (m : SparseMatrix R) (p : Nat × Nat) : Nat × Nat × R :=
+  let ((row, col), val) := Option.getD (Std.RBMap.findEntry? m.entries p) ((0,0),0)
+  (row, col, val)
+
+/--
+Returns an element by indices, whenever it exists, or 0 otherwise
+-/
+def SparseMatrix.findElem (m : SparseMatrix R) (p : Nat × Nat) : R :=
+  (SparseMatrix.findEntry m p).2.2
+
+/--
+Sparse matrix addition
+-/
+def SparseMatrix.addition
+  (m₁ : SparseMatrix R) (m₂ : SparseMatrix R) : SparseMatrix R := Id.run do
+  let cij i j := findD m₁.entries (i,j) 0 + findD m₂.entries (i,j) 0
+  let (cols, rows) := (m₁.cols, m₁.rows)
+  if dim m₁ == dim m₂ then
+    let mut acc : SparseMatrix R := SparseMatrix.empty cols rows
+    for i in [0:rows] do
+      for j in [0:cols] do
+        acc := SparseMatrix.insert i j (cij i j) acc
```

Comment:
This logic will fill the sparse matrix with 0's, which is suboptimal. We want to have as little data as possible.

We need some kind of paired match with `find?`. For addition, the logic is:
- If we find elements on both matrices, we insert their sum
- If we only find one element and it's 0, we do nothing
- If we only find one element, insert it
- If we don't find any element, do nothing.

The general idea is that the resulting matrix will have no data other than non-zero elements

Reply:
Let me see